# Models (PyTorch)

Install the Transformers, Datasets, and Evaluate libraries to run this notebook.

In [1]:
!pip install datasets evaluate transformers[sentencepiece]

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 5.0 MB/s eta 0:00:00


In [2]:
from transformers import BertConfig, BertModel

# Building the config
config = BertConfig()

# Building the model from the config
model = BertModel(config)

In [3]:
print(config)

BertConfig {
  "add_cross_attention": false,
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": null,
  "classifier_dropout": null,
  "eos_token_id": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "is_decoder": false,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "tie_word_embeddings": true,
  "transformers_version": "5.0.0",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}



In [4]:
from transformers import BertConfig, BertModel

config = BertConfig()
model = BertModel(config)

# Model is randomly initialized!

In [5]:
from transformers import BertModel

model = BertModel.from_pretrained("bert-base-cased")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:104: UserWarning: 
Error while fetching `HF_TOKEN` secret value from your vault: 'Requesting secret HF_TOKEN timed out. Secrets can only be fetched when running from the Colab UI.'.
You are not authenticated with the Hugging Face Hub in this notebook.
If the error persists, please let us know by opening an issue on GitHub (https://github.com/huggingface/huggingface_hub/issues/new).
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/436M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-cased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [6]:
model.save_pretrained("directory_on_my_computer")

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

In [7]:
sequences = ["Hello!", "Cool.", "Nice!"]

In [8]:
encoded_sequences = [
    [101, 7592, 999, 102],
    [101, 4658, 1012, 102],
    [101, 3835, 999, 102],
]

In [9]:
import torch

model_inputs = torch.tensor(encoded_sequences)

In [10]:
output = model(model_inputs)

In [11]:
output

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[ 4.4496e-01,  4.8276e-01,  2.7797e-01,  ..., -5.4033e-02,
           3.9394e-01, -9.4770e-02],
         [ 2.4943e-01, -4.4093e-01,  8.1772e-01,  ..., -3.1917e-01,
           2.2992e-01, -4.1172e-02],
         [ 1.3668e-01,  2.2518e-01,  1.4502e-01,  ..., -4.6915e-02,
           2.8224e-01,  7.5566e-02],
         [ 1.1789e+00,  1.6738e-01, -1.8187e-01,  ...,  2.4671e-01,
           1.0441e+00, -6.1958e-03]],

        [[ 3.6436e-01,  3.2464e-02,  2.0258e-01,  ...,  6.0111e-02,
           3.2451e-01, -2.0996e-02],
         [ 7.1866e-01, -4.8725e-01,  5.1740e-01,  ..., -4.4012e-01,
           1.4553e-01, -3.7545e-02],
         [ 3.3223e-01, -2.3271e-01,  9.4876e-02,  ..., -2.5268e-01,
           3.2172e-01,  8.1129e-04],
         [ 1.2523e+00,  3.5754e-01, -5.1320e-02,  ..., -3.7840e-01,
           1.0526e+00, -5.6255e-01]],

        [[ 2.4042e-01,  1.4718e-01,  1.2110e-01,  ...,  7.6062e-02,
           3.3564e-01,  2

In [12]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

encoded_input = tokenizer("Hello, I'm a single sentence!")
print(encoded_input)

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

{'input_ids': [101, 8667, 117, 146, 112, 182, 170, 1423, 5650, 106, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}


In [13]:
encoded_input

{'input_ids': [101, 8667, 117, 146, 112, 182, 170, 1423, 5650, 106, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [14]:
print(tokenizer.decode(encoded_input["input_ids"]))

[CLS] Hello, I ' m a single sentence! [SEP]


In [15]:
encoded_input = tokenizer(
    ["How are you?", "I'm fine, thank you!"],
    padding=True,
    truncation=True,
    max_length=5,
    return_tensors="pt",
)
print(encoded_input)

{'input_ids': tensor([[ 101, 1731, 1132, 1128,  102],
        [ 101,  146,  112,  182,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]])}


### Just run a Pipeline test -> Tokenizer -> Embedding ID -> Convert into Tensor -> Pass Into Model.

In [16]:
sequences = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]

In [17]:
encoded_sequences = tokenizer(
    sequences,
    padding=True,
    truncation=True,
    max_length=5,
    return_tensors="pt",
)

In [18]:
encoded_sequences

{'input_ids': tensor([[ 101,  146,  112, 1396,  102],
        [ 101,  146, 4819, 1142,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]])}

In [19]:
from transformers import BertModel

model = BertModel.from_pretrained("bert-base-cased")

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: bert-base-cased
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.bias                       | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [20]:
output = model(**encoded_sequences)

In [21]:
output

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[ 0.5733,  0.1890,  0.2006,  ..., -0.1063,  0.3804, -0.2661],
         [ 0.6559,  0.0920,  0.6745,  ..., -0.0041,  0.1934,  0.1391],
         [ 0.3307,  0.3666,  0.0287,  ..., -0.0191,  0.9605, -0.2183],
         [ 0.3565, -0.1648,  0.0481,  ..., -0.1992, -0.1487,  0.0479],
         [ 1.0243,  0.4982, -0.2137,  ..., -0.4886,  0.7086, -1.0326]],

        [[ 0.5044,  0.4906,  0.0981,  ..., -0.3185,  0.3671, -0.4463],
         [ 0.7304,  0.0368,  0.4456,  ..., -0.1137,  0.3073,  0.3212],
         [ 0.2741,  0.1903, -0.2275,  ...,  0.4166,  0.0118,  0.4595],
         [ 0.0973, -0.3476, -0.0465,  ..., -0.4456,  0.4295,  0.1949],
         [ 0.8705,  0.6041,  0.0960,  ..., -0.1663,  0.3639, -0.5766]]],
       grad_fn=<NativeLayerNormBackward0>), pooler_output=tensor([[-0.6450,  0.3815,  0.9998,  ...,  0.9999, -0.8419,  0.9852],
        [-0.5711,  0.4728,  0.9998,  ...,  1.0000, -0.9377,  0.9948]],
       grad_fn=<TanhBack

#### The output you're seeing, BaseModelOutputWithPoolingAndCrossAttentions, is a standard output format from the Hugging Face transformers library for models like BERT. Let's break down its key components:

last_hidden_state: This is a tensor representing the sequence of hidden states generated by the last layer of the BERT model. Each row in this tensor corresponds to an input token (including special tokens like [CLS] and [SEP]). The values in each row are a 768-dimensional vector (since your BERT model has a hidden_size of 768), which is the contextualized embedding for that token. This output is extremely useful for tasks that require per-token understanding, such as named entity recognition or question answering.

pooler_output: This is a tensor that typically represents the entire input sequence. Specifically, it's the hidden state corresponding to the [CLS] token (the first token in each sequence), after it has been passed through a pooling layer (often a linear layer followed by a Tanh activation function). This pooler_output is often used for sentence-level classification tasks, as it's designed to capture a summary representation of the entire sequence.

hidden_states, past_key_values, attentions, cross_attentions: These are None by default in your output. They would contain additional, more granular information about the model's internal workings (e.g., the outputs of all hidden layers, or the attention weights), if you had configured the model to produce them by setting parameters like output_hidden_states=True or output_attentions=True when loading the model.

What the model does with an encoded sequence as input:

When you provide an encoded_sequence (which contains input_ids, attention_mask, and token_type_ids) to a BERT model:

Embeddings: The model first converts the input_ids (numerical representations of tokens) into dense vector representations called embeddings. This involves combining three types of embeddings:

Token Embeddings: A unique vector for each word/subword token in your vocabulary.
Position Embeddings: Vectors that encode the position of each token in the sequence, as Transformer models inherently don't understand word order.
Segment Embeddings: Vectors that indicate which sentence a token belongs to (useful for tasks involving two sentences, like natural language inference).
Transformer Layers: These combined embeddings are then fed through multiple (in your case, 12) identical Transformer encoder layers. Each layer consists of two main sub-layers:

Multi-Head Self-Attention: This mechanism allows the model to weigh the importance of all other tokens in the sequence when processing a specific token. This is how BERT understands context and relationships between words.
Feed-Forward Network: A simple neural network applied independently to each position.
Output: After passing through all the Transformer layers, the model produces the last_hidden_state (the contextualized embeddings for each token at the final layer) and optionally the pooler_output (a consolidated representation for the entire sequence, usually derived from the [CLS] token). These outputs are then used as features for downstream tasks like classification, question answering, or text generation.

In essence, the model processes your input tokens, enriches them with positional and segment information, and then iteratively refines their representations by allowing them to 'attend' to each other, ultimately producing highly contextualized embeddings that capture the meaning and relationships within your text.

In [22]:
for i, input_ids in enumerate(encoded_sequences['input_ids']):
    decoded_text = tokenizer.decode(input_ids)
    print(f"Decoded sequence {i+1}: {decoded_text}")

Decoded sequence 1: [CLS] I ' ve [SEP]
Decoded sequence 2: [CLS] I hate this [SEP]


{
  "add_cross_attention": false,
  "architectures": [
    "BertModel"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": null,
  "classifier_dropout": null,
  "dtype": "float32",
  "eos_token_id": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "is_decoder": false,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "tie_word_embeddings": true,
  "transformers_version": "5.0.0",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 28996
}


## This config.json file provides the architectural details and hyperparameters of your BERT model. Let's break down each piece of information:

add_cross_attention: false. This parameter is usually true for decoder models in an encoder-decoder setup (like in a T5 or BART model), indicating that the decoder can attend to the encoder's output. For an encoder-only model like BERT, it's typically false.

architectures: ["BertModel"]. This lists the specific model classes that this configuration is designed for. Here, it confirms it's a BERT-style model.

attention_probs_dropout_prob: 0.1. The dropout probability applied to the attention probabilities (weights) inside the self-attention layers. This helps prevent overfitting.

bos_token_id: null. The ID of the Beginning Of Sentence (BOS) token. BERT doesn't typically use a separate BOS token, relying on [CLS] for sequence representation.

classifier_dropout: null. Dropout probability for the classifier head if the model is used for classification tasks. It can be set when fine-tuning.

dtype: "float32". The data type used for the model's parameters.

eos_token_id: null. The ID of the End Of Sentence (EOS) token. BERT uses [SEP] for this purpose, not a dedicated EOS token.

gradient_checkpointing: false. A technique to save memory during training by recomputing some activations instead of storing them, at the cost of increased computation time.

hidden_act: "gelu". The activation function used in the hidden layers, specifically in the feed-forward networks within each Transformer block.

hidden_dropout_prob: 0.1. The dropout probability applied to the outputs of all sub-layers (attention and feed-forward) in the encoder.

hidden_size: 768. This is a crucial parameter, representing the dimensionality of the embeddings and the hidden states throughout the model. Every token in your sequence will be represented by a 768-dimensional vector after passing through the model's layers.

initializer_range: 0.02. The standard deviation of the truncated normal initializer used to initialize all weight matrices.

intermediate_size: 3072. This is the size of the 'intermediate' or 'feed-forward' layer within each Transformer block. It's typically 4 times the hidden_size (4 * 768 = 3072 in this case).

is_decoder: false. This is key! It indicates whether the model is a decoder. For BERT, which is an encoder-only model, this is false.

layer_norm_eps: 1e-12. The epsilon value added to the denominator of the Layer Normalization computation for numerical stability.

max_position_embeddings: 512. The maximum sequence length that this model can handle. If your input sequence is longer than this, it will be truncated.

model_type: "bert". This specifies the overall architecture family of the model.

num_attention_heads: 12. The number of attention heads in each multi-head attention layer. The hidden_size is divided by this number to get the size of each head (768 / 12 = 64).

num_hidden_layers: 12. The number of Transformer encoder layers in the model.

pad_token_id: 0. The ID of the padding token, used to make sequences of different lengths the same size.

position_embedding_type: "absolute". Indicates how positional information is added to the token embeddings.

tie_word_embeddings: true. Whether the input and output word embeddings should be tied (i.e., share the same weights). Common in models with language generation capabilities.

transformers_version: "5.0.0". The version of the Hugging Face Transformers library used to save this configuration.

type_vocab_size: 2. The size of the token type (segment) vocabulary. BERT uses two types, typically for distinguishing two sentences in tasks like Next Sentence Prediction (0 for sentence A, 1 for sentence B).

use_cache: true. Whether the model should return the last key/values attention states (useful for faster decoding in generation tasks).

vocab_size: 28996. The total number of unique tokens in the model's vocabulary.

Encoder-only vs. Decoder-only vs. Encoder-Decoder Models:
You to identify the type of model by looking at a few key parameters:

model_type: This is a primary indicator. For example, "bert" and "roberta" are typically encoder-only. "gpt2", "llama", "mistral" are decoder-only. "t5", "bart" are encoder-decoder.

is_decoder: This is a direct flag. If true, the model (or this part of an encoder-decoder model) functions as a decoder. If false (as in your BERT config), it's an encoder. An encoder-decoder model will typically have is_decoder: false for its encoder part and is_decoder: true for its decoder part.

add_cross_attention: If is_decoder is true and add_cross_attention is true, it means this decoder is designed to attend to an encoder's output, indicating an encoder-decoder architecture.

Architectural Clues: Encoder-only models are great for tasks like classification, named entity recognition, and question answering. Decoder-only models excel at text generation. Encoder-decoder models are used for sequence-to-sequence tasks like translation and summarization.

Based on your config with model_type: "bert" and is_decoder: false, this model is definitively an encoder-only architecture.

What is a Dense Layer?
In the context of this config.json and Transformer models, a "dense layer" typically refers to a fully connected layer in a neural network, also known as a linear layer. It's a layer where every input neuron is connected to every output neuron, each connection having its own weight.

Specifically, within the Transformer architecture, you'll find dense layers in two main places:

Feed-Forward Networks (FFN): Each Transformer block contains a feed-forward network. This network usually consists of two dense layers with an activation function (like GELU, specified by hidden_act) in between. The intermediate_size (3072 in your config) refers to the size of the hidden dimension in the first dense layer of this FFN, before it's projected back to the hidden_size (768). So, it expands the dimensionality and then contracts it.
Pooling Layers: The pooler_output you saw earlier comes from a dense layer applied to the [CLS] token's hidden state.
Output Layers (Heads): If you were to fine-tune this model for a specific task (e.g., classification), you would typically add a task-specific head on top of the last_hidden_state or pooler_output, and this head would often start with one or more dense layers.
So, when your config mentions intermediate_size, it's directly referencing the size of a key dense layer within each of the num_hidden_layers Transformer blocks.